### PR TITLE
fix(hydroflow_lang): allow `ensure_singleton_referencers_succeed_persist` to ignore `identity`/`tee`/`union` operators, helps #1290

### PR DESCRIPTION
**Pull request by MingweiSamuel**
Adds tests for `ensure_singleton_referencers_succeed_persist` persist insertion behavior

Fixes some cases of #1290, when just identity operators (union/tee/identity) are between the singleton referencer and its preceding flow state.

### CODE REVIEW

**Comment by MingweiSamuel** [comment]
Need to check that this behaves correctly with sources

File: hydroflow_lang/src/graph/flat_graph_builder.rs [hunk 1]
```diff
@@ -902,18 +902,42 @@ impl FlatGraphBuilder {
...
                 .flat_graph
                 .node_predecessors(node_id)
                 .collect::<Vec<_>>()
             {
-                if let Some(op_inst) = self.flat_graph.node_op_inst(pred_node_id) {
-                    if "persist" == op_inst.op_constraints.name
-                        || "state" == op_inst.op_constraints.name
-                    {
-                        continue;
+                // Check if the predacessor is already persisted. If so, no persist should be inserted.
```
Comment (rohitkulshreshtha) [inline hydroflow_lang/src/graph/flat_graph_builder.rs:911]: Typo: s/predacessor/predecessor/.

Comment (MingweiSamuel) [reply hydroflow_lang/src/graph/flat_graph_builder.rs:911]: Ah I always spell that word wrong lol

File: hydroflow_lang/src/graph/flat_graph_builder.rs [hunk 2]
```diff
@@ -902,18 +902,42 @@ impl FlatGraphBuilder {
...
+                        };
+                        match op_inst.op_constraints.name {
+                            // Stateful operators, we're good.
+                            "persist" | "state" => {}
+                            // Identity operators, check transitively.
+                            // TODO(mingwei): Add `transparent` or `is_identity` field to
+                            // `OperatorInstance` so we don't have to list all variants here.
+                            "identity" | "union" | "tee" => {
+                                pred_node_ids
+                                    .extend(self.flat_graph.node_predecessor_nodes(pred_node_id));
```
Comment (rohitkulshreshtha) [inline hydroflow_lang/src/graph/flat_graph_builder.rs:928]: Should we turn the responsibility around to the operators? The operator should tell this part of the code what interests it.  

In other words. Should there be an `operator.persisted_at_end()` or something of that nature, this code only checks if that is true.

Comment (MingweiSamuel) [reply hydroflow_lang/src/graph/flat_graph_builder.rs:928]: Yeah had the TODO for it, I will make it a separate issue

Comment (MingweiSamuel) [reply hydroflow_lang/src/graph/flat_graph_builder.rs:928]: https://github.com/hydro-project/hydroflow/issues/1311

**Review by rohitkulshreshtha — APPROVED** [review]

**Pull request merged**